README.md
# 獅尾肉丸 Swei Meatball

獅尾肉丸字體基於[思源宋體](https://github.com/adobe-fonts/source-han-serif)的三角形換半圓形改造，更加簡明現代化的字體。支援简体中文、繁體中文、韓文與日文；歡迎大家自由應用、自由優化、自由改作！

相比於思源宋體，獅尾肉丸的造型更加簡明現代，版面效果清新輕快。

「思源宋體」、「獅尾四季春」、「獅尾肉丸」的比較：
![字體比較](https://github.com/max32002/swei-meatball/raw/master/preview/compare_noto_serif.png)

和思源宋體一樣，支援7種的字重：
![字重比較](https://github.com/max32002/swei-meatball/raw/master/preview/compare_style.png)

* 在「肉」、「糸」、「女」、「辶」、「食」的筆畫，獅尾比較相似台灣教育部推薦字體筆順，適合教育用途。
* 在「草」部的筆畫，獅尾是分開的二個部份。
* 字體裡的「草字頭部首」中間該不該連起來，「肉字旁部首」該不該變成「月」，要不要把手寫習慣代入印刷字體？整體看來，會不會影響視覺的延伸性？會不會影響印刷的可行性與閱讀的便利性？這個答案我也不清楚，如果你是台灣教育部標準字符的愛好者，獅尾字體應該是一個不錯的選擇。

## 更新日誌
[點擊此處](https://github.com/max32002/swei-meatball/blob/master/change_log.md) 查看更新記錄。

## 下載字型

請點選GitHub此畫面右上綠色「Clone or download」按鈕，並選擇「Download ZIP」，或點進想下載的ttf字型檔案，再點「Download」的按鈕進行下載。

## 附註

* 筆畫太複雜或重疊的，因為拆筆畫有點難，目前是先跳過，所以這不是一個完整、健康和專業的字體。
* 由於小編對字體編碼方式完全不清楚，相較於原版的思源宋體，可能有掉很多符號或不常用的字。
* 演算黑科技將字體三角形換成圓形，請參考 /python/ 目錄下的腳本檔案。透過調整程式碼，也許也可以產生出新的有趣字型。目前的程式應該還有很多錯誤或需要再加強的地方。請先把要處理的字型，轉成 FontForge 的 sfdir, 再透過程式去處理產生出來的檔案。

## 著作權與授權

* 本字型是基於 SIL Open Font License 1.1 改造Adobe和Google所開發、發表的「[思源宋體](https://github.com/adobe-fonts/source-han-serif)」字型。
* 本字型亦基於 SIL Open Font License 1.1 授權條款免費公開，關於授權合約的內容、免責事項等細節，請詳讀 License 文件。
    * 可自由商用 不需付費、知會或標明作者，即可自由使用此字型，亦可做商業應用。
    * 可自由傳布 可自由分享檔案、將檔案安裝於任何軟硬體中。
    * 可自由改作為其他字型 將字型檔案修改重製為其他字型檔案，改作後的字型檔案須同樣依 Open Font License 釋出。
    
    
## 相關網頁

* 台灣圓體 TaiwanPearl
https://max-everyday.com/2020/06/taiwanpearl/
* 獅尾肉丸 Swei Meatball
https://max-everyday.com/2020/06/swei-meatball/
* 獅尾四季春字體 Swei Spring
https://max-everyday.com/2020/04/swei-spring/
* 獅尾半月字體 Swei Gothic
https://max-everyday.com/2020/04/swei-half-moon/
* 獅尾圓體 Swei Gothic
https://max-everyday.com/2020/04/swei-gothic/
* 獅尾黑體 Swei Sans
https://max-everyday.com/2020/03/swei-sans/
* 內海字體 (NaikaiFont) 
https://max-everyday.com/2020/03/naikaifont/
* 莫大毛筆字體 (Bakudai)
https://max-everyday.com/2020/03/bakudaifont/
* 正風毛筆字體 (MasaFont)
https://max-everyday.com/2020/05/masafont/
* 清松手寫體 (JasonHandWriting)
https://jasonfonts.max-everyday.com/
* 源泉圓體
https://github.com/ButTaiwan/gensen-font
* jf open 粉圓
https://github.com/justfont/open-huninn-font
* 假粉圓體
https://github.com/max32002/FakePearl
* 俊羽圓體
https://github.com/max32002/YuPearl

## 贊助Max

很高興可以替中華民國美學盡一分心力，如果你覺得這篇文章寫的很好，想打賞Max，贊助方式如下：
https://max-everyday.com/about/#donate


change_log.md
# 2020/06/16 ver 1.042
* 針對部份筆畫重疊的三角形，轉成圓形。

# 2020/06/16 ver 1.040
* 第1版。

python/TtfConfig.py
#!/usr/bin/env python3
#encoding=utf-8

class TtfConfig():
    VERSION = "1.047"
    PROCESS_MODE = "SPRING"
    PROCESS_MODE = "MEATBALL"

    STYLE_INDEX = 4
    STYLE_ARRAY = ["Black","Bold","SemiBold","Medium","Regular","Light","ExtraLight"]
    STYLE=STYLE_ARRAY[STYLE_INDEX]

    DEFAULT_COORDINATE_VALUE = -9999

    # unicode in field
    # 1 to 3
    UNICODE_FIELD = 2

    #BMP_PATH = '/Users/chunyuyao/Documents/noto/bmp'
    # don't use bmp feature.
    BMP_PATH = None

    # for X,Y axis equal compare.
    # each 100 px, +- 8 px.
    EQUAL_ACCURACY_MIN = 3
    EQUAL_ACCURACY_PERCENT = 0.08

    # for rule#101
    # in regular=80 to 57
    # for case:.2639 「真」被擠壓為38
    # for case:uni7665 癥，中間的「山」為29
    ROW_TRIANGLE_HEIGHT_MIN = 26
    # for case:uni4E00 一：104
    ROW_TRIANGLE_HEIGHT_MAX = 115
    
    # in regular=121 to 86
    # for uni5A64「婤」, value:63, 因為被截斷
    # for uni8061,聡: 48
    ROW_TRIANGLE_SLIDE_MIN = 44
    # for case:uni4E00 一：154
    ROW_TRIANGLE_SLIDE_MAX = 165

    # in regular=29,31 to (.26356, 片)47, (case.10924 公, Regular) 73
    ROW_TRIANGLE_CHIN_MIN = 21
    # in regular=85 (uni6C43 汃)
    ROW_TRIANGLE_CHIN_MAX = 95

    # in regular=46, (case.11158 分, Regular) 69
    ROW_TRIANGLE_FLAT_CHIN_MIN = 26
    ROW_TRIANGLE_FLAT_CHIN_MAX = 84

    # for rule#101
    # in regular=103 to 112
    STROKE_WIDTH_MIN = 43
    STROKE_WIDTH_MAX = 133

    # in regular=45 to 42, (case:.17729, 後) 34, (case:.18791, 應) 26
    # for uni8763 蝣的子，19
    COL_TRIANGLE_CHIN_MIN = 17
    COL_TRIANGLE_CHIN_MAX = 55

    def apply_weight_setting(self):
        self.STYLE=self.STYLE_ARRAY[self.STYLE_INDEX]

        if self.STYLE in ["Black"]:
            # black style, for case:三, 131
            self.ROW_TRIANGLE_HEIGHT_MIN = 28
            self.ROW_TRIANGLE_HEIGHT_MAX = 151

            # black style, for case:三, 186
            self.ROW_TRIANGLE_SLIDE_MIN = 54
            self.ROW_TRIANGLE_SLIDE_MAX = 200

            # black style, for case:三, 34, (case.10924 公, Black) 104
            self.ROW_TRIANGLE_CHIN_MIN = 21
            self.ROW_TRIANGLE_CHIN_MAX = 125

            # in regular=46
            self.ROW_TRIANGLE_FLAT_CHIN_MIN = 26
            self.ROW_TRIANGLE_FLAT_CHIN_MAX = 146

            # for rule#101
            # in regular=103 to 112, (case.10924 公, Black) 216
            self.STROKE_WIDTH_MIN = 63
            self.STROKE_WIDTH_MAX = 231
            
            # in regular=45 to 42
            self.COL_TRIANGLE_CHIN_MIN = 25
            self.COL_TRIANGLE_CHIN_MAX = 85

        if self.STYLE in ["Bold"]:
            # in regular=80 to 57
            self.ROW_TRIANGLE_HEIGHT_MIN = 28
            self.ROW_TRIANGLE_HEIGHT_MAX = 141
            # in regular=121 to 86
            self.ROW_TRIANGLE_SLIDE_MIN = 52
            self.ROW_TRIANGLE_SLIDE_MAX = 190
            # in regular=29,31 to (.26356, 片)47
            self.ROW_TRIANGLE_CHIN_MIN = 21
            self.ROW_TRIANGLE_CHIN_MAX = 116

            # in regular=46
            self.ROW_TRIANGLE_FLAT_CHIN_MIN = 26
            self.ROW_TRIANGLE_FLAT_CHIN_MAX = 126

            # for rule#101
            # in regular=103 to 112, (case.10924 公, Black) 216
            self.STROKE_WIDTH_MIN = 58
            self.STROKE_WIDTH_MAX = 208

            # in regular=45 to 42
            self.COL_TRIANGLE_CHIN_MIN = 23
            self.COL_TRIANGLE_CHIN_MAX = 85


        if self.STYLE in ["SemiBold","Medium"]:
            # in regular=80 to 57
            self.ROW_TRIANGLE_HEIGHT_MIN = 28
            self.ROW_TRIANGLE_HEIGHT_MAX = 130
            # in regular=121 to 86
            self.ROW_TRIANGLE_SLIDE_MIN = 46
            self.ROW_TRIANGLE_SLIDE_MAX = 180
            # in regular=29,31 to (.26356, 片)47
            self.ROW_TRIANGLE_CHIN_MIN = 21
            self.ROW_TRIANGLE_CHIN_MAX = 106

            # in regular=46
            self.ROW_TRIANGLE_FLAT_CHIN_MIN = 26
            self.ROW_TRIANGLE_FLAT_CHIN_MAX = 106

            # for rule#101
            # in regular=103 to 112
            self.STROKE_WIDTH_MIN = 53
            self.STROKE_WIDTH_MAX = 183

            # in regular=45 to 42
            self.COL_TRIANGLE_CHIN_MIN = 20
            self.COL_TRIANGLE_CHIN_MAX = 75

        if self.STYLE in ["Light","ExtraLight"]:
            # in regular=80 to 57
            self.ROW_TRIANGLE_HEIGHT_MIN = 24
            self.ROW_TRIANGLE_HEIGHT_MAX = 110
            # in regular=121 to 86
            self.ROW_TRIANGLE_SLIDE_MIN = 39
            self.ROW_TRIANGLE_SLIDE_MAX = 160
            # in regular=29,31 to (.26356, 片)47
            self.ROW_TRIANGLE_CHIN_MIN = 18
            self.ROW_TRIANGLE_CHIN_MAX = 88

            # in regular=46
            self.ROW_TRIANGLE_FLAT_CHIN_MIN = 22
            self.ROW_TRIANGLE_FLAT_CHIN_MAX = 80

            # for rule#101
            # in regular=103 to 112
            self.STROKE_WIDTH_MIN = 38
            self.STROKE_WIDTH_MAX = 133

            # in regular=45 to 42
            self.COL_TRIANGLE_CHIN_MIN = 17
            self.COL_TRIANGLE_CHIN_MAX = 55

    def __init__(self, weight_code):
        import datetime

        self.STYLE_INDEX = int(weight_code)
        self.apply_weight_setting()
        print("Transform Mode:", self.PROCESS_MODE)
        print("Transform Style:", self.STYLE)
        print("Transform Version:", self.VERSION)
        print("Transform Time:", datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"))

    def hello(self):
        print("world!")

python/util/Rule102_Col.py
#!/usr/bin/env python3
#encoding=utf-8

from . import spline_util
from . import Rule

# RULE # 102
# 直線開頭三角形轉長方形
# PS: 因為 array size change, so need redo.
class Rule(Rule.Rule):
    def __init__(self):
        pass

    def apply(self, spline_dict, resume_idx, inside_stroke_dict,skip_coordinate, skip_coordinate_rule):
        redo_travel=False
        check_first_point = False

        # default（normal): 1.3 to 1.43, (橫線) 1.733, (for case:.26335 狀) 1.233
        SLIDE_1_PERCENT_MIN = 1.11
        SLIDE_1_PERCENT_MAX = 1.83

        # default: 0.7 to 0.53, 
        SLIDE_2_PERCENT_MIN = 0.43
        SLIDE_2_PERCENT_MAX = 0.8

        # default: 1.76 to 1.9, (橫線) 1.9-1.96, (for case:.26335 狀) 1.633
        # for case uni6507 攇, 1.99
        SLIDE_3_PERCENT_MIN = 1.52
        SLIDE_3_PERCENT_MAX = 1.99

        # for uni83A3 莣, 1.25/0.81/1.79
        SLIDE_11_PERCENT_MIN = 1.05
        SLIDE_11_PERCENT_MAX = 1.45
        SLIDE_12_PERCENT_MIN = 0.50
        SLIDE_12_PERCENT_MAX = 1.01
        SLIDE_13_PERCENT_MIN = 1.59
        SLIDE_13_PERCENT_MAX = 1.99

        # for case: uni8F9C 辜的辛的立的右上角。 0.325/0.725/1.98
        # for case: uni9059 遙的缶。0.865 / 0.6 / 1.95
        SLIDE_21_PERCENT_MIN = 0.00
        SLIDE_21_PERCENT_MAX = 1.45

        # clone
        format_dict_array=[]
        format_dict_array = spline_dict['dots'][1:]
        format_dict_array = self.caculate_distance(format_dict_array)

        nodes_length = len(format_dict_array)
        #print("orig nodes_length:", len(spline_dict['dots']))
        #print("format nodes_length:", nodes_length)
        #print("resume_idx:", resume_idx)

        rule_need_lines = 4
        fail_code = -1
        if nodes_length >= rule_need_lines:
            for idx in range(nodes_length):
                if idx <= resume_idx:
                    # skip traveled nodes.
                    continue

                is_debug_mode = False
                #is_debug_mode = True

                # 要轉換的角，不能就是我們產生出來的點。
                if [format_dict_array[idx]['x'],format_dict_array[idx]['y']] in skip_coordinate:
                    if is_debug_mode:
                        print("match skip dot +0:",[format_dict_array[(idx+0)%nodes_length]['x'],format_dict_array[(idx+0)%nodes_length]['y']])
                        pass
                    continue

                if format_dict_array[idx]['code'] in skip_coordinate_rule:
                    if is_debug_mode:
                        print("match skip skip_coordinate_rule +0:",[format_dict_array[(idx+1)%nodes_length]['x'],format_dict_array[(idx+1)%nodes_length]['y']])
                        pass
                    continue

                is_debug_mode = False
                #is_debug_mode = True

                if is_debug_mode:
                    debug_coordinate_list = [[186,83]]
                    if not([format_dict_array[idx]['x'],format_dict_array[idx]['y']] in debug_coordinate_list):
                        continue

                    print("="*30)
                    print("index:", idx)
                    for debug_idx in range(8):
                        print(debug_idx-2,": val#102:",format_dict_array[(idx+debug_idx+nodes_length-2)%nodes_length]['code'],'-(',format_dict_array[(idx+debug_idx+nodes_length-2)%nodes_length]['distance'],')')

                # begin travel.
                is_match_pattern = True

                # match ??lc?
                if is_match_pattern:
                    fail_code = 100
                    is_match_pattern = False
                    if format_dict_array[(idx+2)%nodes_length]['t'] == 'l':
                        if format_dict_array[(idx+3)%nodes_length]['t'] == 'c':
                            is_match_pattern = True

                is_more_one_dot = False

                if is_match_pattern:
                    fail_code = 200
                    is_match_pattern = False
                    if format_dict_array[(idx+1)%nodes_length]['distance'] >= self.config.COL_STROKE_WIDTH_MIN:
                        fail_code = 210
                        if format_dict_array[(idx+1)%nodes_length]['distance'] <= self.config.COL_STROKE_WIDTH_MAX:
                            fail_code = 220
                            
                            if format_dict_array[(idx+2)%nodes_length]['distance'] >= self.config.COL_TRIANGLE_CHIN_MIN:
                                fail_code = 230
                                if format_dict_array[(idx+2)%nodes_length]['distance'] <= self.config.COL_TRIANGLE_CHIN_MAX:
                                    is_match_pattern = True

                            # +2 & +3 are small edge.
                            # for case: uni6AB9 檹
                            # ps: 可能同時 is_match_pattern = True 在 fail_code 230 還有下面的 fail_code 240
                            small_edge_sum = format_dict_array[(idx+2)%nodes_length]['distance']+format_dict_array[(idx+3)%nodes_length]['distance']
                            #print("small_edge_sum:",small_edge_sum)
                            if format_dict_array[(idx+2)%nodes_length]['distance'] < 45:
                                if format_dict_array[(idx+3)%nodes_length]['distance'] < 45:
                                    fail_code = 240
                                    if small_edge_sum >= self.config.COL_TRIANGLE_CHIN_MIN:
                                        if small_edge_sum <= self.config.COL_TRIANGLE_CHIN_MAX:
                                            if format_dict_array[(idx+2)%nodes_length]['x_direction'] == format_dict_array[(idx+3)%nodes_length]['x_direction']:
                                                if format_dict_array[(idx+2)%nodes_length]['y_direction'] == format_dict_array[(idx+3)%nodes_length]['y_direction']:
                                                    is_more_one_dot = True
                                                    is_match_pattern = True

                # for case: uni8F9C 辜的辛的立的右上角。
                # 是否第一個邊，與其他筆畫合併。
                is_first_edge_merged = False

                if is_match_pattern:
                    fail_code = 300
                    is_match_pattern = False
                    # 「/」頭在上。
                    if format_dict_array[(idx+0)%nodes_length]['x_direction'] >= 0:
                        fail_code = 310
                        if format_dict_array[(idx+1)%nodes_length]['x_direction'] > 0:
                            fail_code = 311
                            if format_dict_array[(idx+2)%nodes_length]['x_direction'] < 0:
                                fail_code = 312
                                if format_dict_array[(idx+3)%nodes_length]['x_direction'] <= 0:
                                    is_match_pattern = True

                    # for case: uni9059 遙的辶的最左下角。
                    # 「/」頭在下。
                    # PS: 目前未拔除。
                    if format_dict_array[(idx+0)%nodes_length]['x_direction'] <= 0:
                        fail_code = 315
                        if format_dict_array[(idx+1)%nodes_length]['x_direction'] > 0:
                            fail_code = 316
                            if format_dict_array[(idx+2)%nodes_length]['x_direction'] < 0:
                                fail_code = 317
                                if format_dict_array[(idx+3)%nodes_length]['x_direction'] >= 0:
                                    is_match_pattern = True

                    # 「＼」
                    if not is_match_pattern:
                        if format_dict_array[(idx+0)%nodes_length]['x_direction'] <= 0:
                            fail_code = 320
                            if format_dict_array[(idx+1)%nodes_length]['x_direction'] > 0:
                                fail_code = 321
                                if format_dict_array[(idx+2)%nodes_length]['x_direction'] < 0:
                                    fail_code = 322
                                    if format_dict_array[(idx+3)%nodes_length]['x_direction'] >= 0:
                                        is_match_pattern = True

                    # 「|」common case 頭在上。
                    if not is_match_pattern:
                        if format_dict_array[(idx+0)%nodes_length]['x_equal_fuzzy']:
                            fail_code = 330
                            if format_dict_array[(idx+1)%nodes_length]['x_direction'] > 0:
                                fail_code = 331
                                if format_dict_array[(idx+2)%nodes_length]['x_direction'] < 0:
                                    fail_code = 332
                                    if format_dict_array[(idx+3)%nodes_length]['x_equal_fuzzy']:
                                        is_match_pattern = True

                    # 「一」接近橫線的版本。
                    if not is_match_pattern:
                        if format_dict_array[(idx+0)%nodes_length]['y_equal_fuzzy']:
                            fail_code = 335
                            if format_dict_array[(idx+1)%nodes_length]['x_direction'] > 0:
                                fail_code = 336
                                if format_dict_array[(idx+2)%nodes_length]['x_direction'] < 0:
                                    fail_code = 337
                                    if format_dict_array[(idx+3)%nodes_length]['y_equal_fuzzy']:
                                        is_match_pattern = True

                    # for case: uni8F9C 辜的辛的立的右上角。
                    # for case: uni9059 遙的缶。
                    # 「/」
                    if not is_match_pattern:
                        if format_dict_array[(idx+0)%nodes_length]['x_direction'] <= 0:
                            fail_code = 340
                            if format_dict_array[(idx+1)%nodes_length]['x_direction'] > 0:
                                fail_code = 341
                                if format_dict_array[(idx+2)%nodes_length]['x_direction'] < 0:
                                    fail_code = 342
                                    if format_dict_array[(idx+3)%nodes_length]['x_direction'] <= 0:
                                        is_match_pattern = True
                                        is_first_edge_merged = True

                if is_debug_mode:
                    print("is_first_edge_merged:", is_first_edge_merged)
                    #print("fail_code:",fail_code)

                if is_match_pattern:
                    fail_code = 400
                    is_match_pattern = False

                    is_pass_dot0_check = False
                    if format_dict_array[(idx+0)%nodes_length]['y_direction'] > 0:
                        fail_code = 410
                        is_pass_dot0_check = True

                    # for case: uni8F9C 辜的辛的立的右上角。
                    if is_first_edge_merged:
                        if format_dict_array[(idx+0)%nodes_length]['y_direction'] <= 0 or format_dict_array[(idx+0)%nodes_length]['y_equal_fuzzy']:
                            fail_code = 411
                            is_pass_dot0_check = True

                    if is_pass_dot0_check:
                        if format_dict_array[(idx+1)%nodes_length]['y_direction'] <= 0:
                            #print(format_dict_array[(idx+2)%nodes_length]['y_direction'])
                            if format_dict_array[(idx+3)%nodes_length]['y1'] <= format_dict_array[(idx+2)%nodes_length]['y']:
                                if format_dict_array[(idx+3)%nodes_length]['y_direction'] < 0:
                                    is_match_pattern = True

                # skip small angle
                if is_match_pattern:
                    fail_code = 500
                    is_match_pattern = False

                    slide_percent_1 = spline_util.slide_percent(format_dict_array[(idx+0)%nodes_length]['x'],format_dict_array[(idx+0)%nodes_length]['y'],format_dict_array[(idx+1)%nodes_length]['x'],format_dict_array[(idx+1)%nodes_length]['y'],format_dict_array[(idx+2)%nodes_length]['x'],format_dict_array[(idx+2)%nodes_length]['y'])
                    slide_percent_2 = spline_util.slide_percent(format_dict_array[(idx+1)%nodes_length]['x'],format_dict_array[(idx+1)%nodes_length]['y'],format_dict_array[(idx+2)%nodes_length]['x'],format_dict_array[(idx+2)%nodes_length]['y'],format_dict_array[(idx+3)%nodes_length]['x'],format_dict_array[(idx+3)%nodes_length]['y'])
                    slide_percent_3 = spline_util.slide_percent(format_dict_array[(idx+2)%nodes_length]['x'],format_dict_array[(idx+2)%nodes_length]['y'],format_dict_array[(idx+3)%nodes_length]['x'],format_dict_array[(idx+3)%nodes_length]['y'],format_dict_array[(idx+4)%nodes_length]['x'],format_dict_array[(idx+4)%nodes_length]['y'])

                    #if True:
                    #if False:
                    if is_debug_mode:
                        print("slide_percent 1:", slide_percent_1)
                        print("data:",format_dict_array[(idx+0)%nodes_length]['x'],format_dict_array[(idx+0)%nodes_length]['y'],format_dict_array[(idx+1)%nodes_length]['x'],format_dict_array[(idx+1)%nodes_length]['y'],format_dict_array[(idx+2)%nodes_length]['x'],format_dict_array[(idx+2)%nodes_length]['y'])
                        print("slide_percent 2:", slide_percent_2)
                        print("data:",format_dict_array[(idx+1)%nodes_length]['x'],format_dict_array[(idx+1)%nodes_length]['y'],format_dict_array[(idx+2)%nodes_length]['x'],format_dict_array[(idx+2)%nodes_length]['y'],format_dict_array[(idx+3)%nodes_length]['x'],format_dict_array[(idx+3)%nodes_length]['y'])
                        print("slide_percent 3:", slide_percent_3)
                        print("data:",format_dict_array[(idx+2)%nodes_length]['x'],format_dict_array[(idx+2)%nodes_length]['y'],format_dict_array[(idx+3)%nodes_length]['x'],format_dict_array[(idx+3)%nodes_length]['y'],format_dict_array[(idx+4)%nodes_length]['x'],format_dict_array[(idx+4)%nodes_length]['y'])

                    # if real line is almost straight, try virtual line.
                    if slide_percent_3 >= 2.0:
                        if format_dict_array[(idx+3)%nodes_length]['t']=='c':
                            x2 = format_dict_array[(idx+3)%nodes_length]['x2']
                            y2 = format_dict_array[(idx+3)%nodes_length]['y2']
                            slide_percent_3 = spline_util.slide_percent(x2,y2,format_dict_array[(idx+3)%nodes_length]['x'],format_dict_array[(idx+3)%nodes_length]['y'],format_dict_array[(idx+4)%nodes_length]['x'],format_dict_array[(idx+4)%nodes_length]['y'])
                            if is_debug_mode:
                                print("slide_percent 3 (x2,y2):", slide_percent_3)
                                print("data:",x2,y2,format_dict_array[(idx+3)%nodes_length]['x'],format_dict_array[(idx+3)%nodes_length]['y'],format_dict_array[(idx+4)%nodes_length]['x'],format_dict_array[(idx+4)%nodes_length]['y'])

                    if slide_percent_1 >= SLIDE_1_PERCENT_MIN and slide_percent_1 <= SLIDE_1_PERCENT_MAX:
                        fail_code = 510
                        if slide_percent_2 >= SLIDE_2_PERCENT_MIN and slide_percent_2 <= SLIDE_2_PERCENT_MAX:
                            fail_code = 520
                            if slide_percent_3 >= SLIDE_3_PERCENT_MIN and slide_percent_3 <= SLIDE_3_PERCENT_MAX:
                                is_match_pattern = True

                    # for uni83A3 莣, 1.25/0.81/1.79
                    if not is_match_pattern:
                        if slide_percent_1 >= SLIDE_11_PERCENT_MIN and slide_percent_1 <= SLIDE_11_PERCENT_MAX:
                            fail_code = 530
                            if slide_percent_2 >= SLIDE_12_PERCENT_MIN and slide_percent_2 <= SLIDE_12_PERCENT_MAX:
                                fail_code = 540
                                if slide_percent_3 >= SLIDE_13_PERCENT_MIN and slide_percent_3 <= SLIDE_13_PERCENT_MAX:
                                    is_match_pattern = True

                    if not is_match_pattern:
                        if is_more_one_dot:
                            slide_percent_2 = spline_util.slide_percent(format_dict_array[(idx+1)%nodes_length]['x'],format_dict_array[(idx+1)%nodes_length]['y'],format_dict_array[(idx+2)%nodes_length]['x'],format_dict_array[(idx+2)%nodes_length]['y'],format_dict_array[(idx+4)%nodes_length]['x'],format_dict_array[(idx+4)%nodes_length]['y'])
                            slide_percent_3 = spline_util.slide_percent(format_dict_array[(idx+2)%nodes_length]['x'],format_dict_array[(idx+2)%nodes_length]['y'],format_dict_array[(idx+4)%nodes_length]['x'],format_dict_array[(idx+4)%nodes_length]['y'],format_dict_array[(idx+5)%nodes_length]['x'],format_dict_array[(idx+5)%nodes_length]['y'])

                            if slide_percent_1 >= SLIDE_1_PERCENT_MIN and slide_percent_1 <= SLIDE_1_PERCENT_MAX:
                                fail_code = 550
                                if slide_percent_2 >= SLIDE_2_PERCENT_MIN and slide_percent_2 <= SLIDE_2_PERCENT_MAX:
                                    fail_code = 551
                                    if slide_percent_3 >= SLIDE_3_PERCENT_MIN and slide_percent_3 <= SLIDE_3_PERCENT_MAX:
                                        is_match_pattern = True

                    # for case: uni8F9C 辜的辛的立的右上角。
                    if not is_match_pattern:
                        if slide_percent_1 >= SLIDE_21_PERCENT_MIN and slide_percent_1 <= SLIDE_21_PERCENT_MAX:
                            fail_code = 560
                            if slide_percent_2 >= SLIDE_12_PERCENT_MIN and slide_percent_2 <= SLIDE_12_PERCENT_MAX:
                                fail_code = 561
                                if slide_percent_3 >= SLIDE_13_PERCENT_MIN and slide_percent_3 <= SLIDE_13_PERCENT_MAX:
                                    is_match_pattern = True

                if is_debug_mode:
                    if not is_match_pattern:
                        print("#", idx,": debug fail_code#101:", fail_code)
                        pass
                    else:
                        print("match rule #102")
                        print(idx,"debug rule102:",format_dict_array[idx]['code'])
                        pass

                if is_match_pattern:
                    if False:
                    #if True:
                        print("#"*40)
                        for debug_idx in range(7):
                            print(debug_idx-2,": values for rule102:",format_dict_array[(idx+debug_idx+nodes_length-2)%nodes_length]['code'],'-(',format_dict_array[(idx+debug_idx+nodes_length-2)%nodes_length]['distance'],')')

                    ending_x = format_dict_array[(idx+3)%nodes_length]['x']
                    #ending_y = format_dict_array[(idx+2)%nodes_length]['y']

                    # update 1
                    old_code = format_dict_array[(idx+2)%nodes_length]['code']
                    old_code_array = old_code.split(' ')
                    #x = int(float(old_code_array[1]))
                    #y = int(float(old_code_array[2]))
                    old_code_array[1]=str(ending_x)
                    new_code = ' '.join(old_code_array)
                    format_dict_array[(idx+2)%nodes_length]['code']=new_code
                    #print("old_code:", old_code)
                    #print("new_code:", new_code)

                    #print("del code:", format_dict_array[(idx+3)%nodes_length]['code'])
                    del format_dict_array[(idx+3)%nodes_length]

                    # we generated nodes
                    #skip_coordinate.append([center_x,center_y])

                    redo_travel=True
                    check_first_point = True
                    resume_idx = -1
                    break

        if check_first_point:
            # check close path.
            self.reset_first_point(format_dict_array, spline_dict)

        return redo_travel, resume_idx, inside_stroke_dict,skip_coordinate, skip_coordinate_rule
